### PR TITLE
feat: 07-14 study

### PR DESCRIPTION
**Pull request by JIWEON-JEONG**

### CODE REVIEW

File: 2024-07-14/컨슈머.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  - ConcurrentKafkaListener 사용 검토하기
+    - spring 추상화 : concurrentKafka
+    - kafka 병령처리 : parallel kafka
+                      
+
+### 컨슈머 배포 전략
+  - 블루 그린 VS 롤링
+    - 기존에 떠있는 서버를 한번에 바꾸느냐, 조금씩 바꾸느냐.
+  - 카나리 & 롤링
+    - V2 로 하나 일단 띄워두고 장애 없으면 배포.
```
Comment (GroovyArea) [inline 2024-07-14/컨슈머.md:64]: 블루 그린의 경우 리밸런싱 1번 일어남,
롤링의 경우 컨슈머와 파티션 개수마다 여러번 일어남 추가 요망

File: 2024-07-14/컨슈머.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+### LAG (병목의 start point 가 queue 일 경우)
+  - 처리 서버 scale out.
+  - 일단 best practice 로 설정해두고 브로커의 리소스 모니터링 하면서 파티션 수 선정
+  - parellel 컨슈밍을 고려해볼수 있다. (cc 면접때 말할거면 제대로 공부하고 가셈)
+
+### 컨슈머 리밸런싱
+  - 컨슈머 그룹 리밸런싱
+    - https://techblog.gccompany.co.kr/카프카-컨슈머-그룹-리밸런싱-kafka-consumer-group-rebalancing-5d3e3b916c9e
+    - 하나의 브로커가 코디네이터 역할을 맡으며, 범위는 컨슈머 그룹이다.
+    - 특정 컨슈머 그룹에서 poll 을 일정시간 하지 않거나 health check
```
Comment (GroovyArea) [inline 2024-07-14/컨슈머.md:30]: heart beat 로 수정

Comment (Hejow) [reply 2024-07-14/컨슈머.md:30]: poll과 commit 둘다~

File: 2024-07-14/컨슈머.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+  - 카프카 모니터링, 관리의 필요성을 느낌.
+  - Confluence 같은 경우 UI 로 모니터링 및 알림 제공.
+
+### 희조님 rabbitMQ (AMQP) 간단 공유 
+  - exchange (brocker), binding(partition - queue 연결), queue (토픽 같은 개념)
+
+### LAG (병목의 start point 가 queue 일 경우)
+  - 처리 서버 scale out.
+  - 일단 best practice 로 설정해두고 브로커의 리소스 모니터링 하면서 파티션 수 선정
+  - parellel 컨슈밍을 고려해볼수 있다. (cc 면접때 말할거면 제대로 공부하고 가셈)
```
Comment (Hejow) [inline 2024-07-14/컨슈머.md:24]: 오타용! parallel

File: 2024-07-14/컨슈머.md (hunk 1 above)
Comment (JIWEON-JEONG) [reply 2024-07-14/컨슈머.md:64]: https://github.com/Be-GGanboo-With-Java/Apache-Kafka-Study/pull/3/commits/fb6c6631546058aeeea93c6bfd235900ab05fd04

File: 2024-07-14/컨슈머.md (hunk 2 above)
Comment (JIWEON-JEONG) [reply 2024-07-14/컨슈머.md]: https://github.com/Be-GGanboo-With-Java/Apache-Kafka-Study/pull/3/commits/caecc78575a93ffa6d8fe818cfbc26c2c9b5497d

File: 2024-07-14/컨슈머.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    이 경우, 컨슈머 A는 `poll()` 메서드를 호출할 때 파티션 0, 1, 2로부터 메시지를 가져옵니다. 컨슈머 B는 파티션 3, 4, 5로부터 메시지를 가져옵니다.
+  
+- 리밸런싱 리스크 
+  - 컨슈머 처리 중단 
+    - 리밸런싱이 완료되기 전에는 컨슈머가 동작 X. (파티션 읽기 작업 중단)
+  - 메시지 중복 처리
+    - Offset Commit 되지 않은 상태에서 리밸런싱 일어날 경우 중복처리 가능하다.
+
+- 해결 방법 
+  - max.poll.records 조정 (500)
```
Comment (Hejow) [inline 2024-07-14/컨슈머.md:46]: records 개수를 줄임으로 해결하는 방법입니다람지~

File: 2024-07-14/컨슈머.md (hunk 3 above)
Comment (JIWEON-JEONG) [reply 2024-07-14/컨슈머.md]: https://github.com/Be-GGanboo-With-Java/Apache-Kafka-Study/pull/3/commits/12c0f2ea06579259c6ffe2ba7d3d307be3232b54

File: 2024-07-14/컨슈머.md (hunk 4 above)
Comment (JIWEON-JEONG) [reply 2024-07-14/컨슈머.md:46]: https://github.com/Be-GGanboo-With-Java/Apache-Kafka-Study/pull/3/commits/e0dfd298b6e2fe01ceeeb42a455454c54a186935

File: 2024-07-14/컨슈머.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    - poll 요청을 통해 consumer 가 Join 하게 되는데, 처리 시간 자체를 줄여 리스크를 줄이는 방법.
```
Comment (Hejow) [inline 2024-07-14/컨슈머.md]:
```suggestion
    - poll 요청을 통해 consumer 가 Join 하게 되는데, 처리해야할 데이터가 적게 만들어 시간 자체를 줄임으로 리스크를 줄이는 방법.
```

**Review by GroovyArea — APPROVED** [review]

**Pull request merged**